prog/weaver/discovery.go

```diff
@@ -0,0 +1,66 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+// TODO: move these definitions somewhere more shareable
+type PeerUpdateRequest struct {
+	Name      string   `json:"peername"`
+	Nickname  string   `json:"nickname"`  // optional
+	Addresses []string `json:"addresses"` // can be empty
+}
+
+type PeerUpdateResponse struct {
+	Addresses []string `json:"addresses"`
+}
+
+var httpClient = &http.Client{Timeout: 30 * time.Second}
+
+func do(verb string, discoveryEndpoint, token string, request interface{}, response interface{}) error {
+	body := new(bytes.Buffer)
+	err := json.NewEncoder(body).Encode(request)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(verb, discoveryEndpoint+"/peer", body)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Scope-Probe token=%s", token))
+	req.Header.Set("X-Weave-Net-Version", version)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
```

Comment:
Just for my understanding, why "`application/x-www-form-urlencoded`" and not "`application/json`" when we actually send JSON?